### DISCUSSION TITLE
Importing STEP geometry and linking it to an IFC Product

### DISCUSSION THREAD

**Discussion by loicgillioz** [IfcOpenShell/IfcOpenShell#3724] [Q&A]
Hello,

I am an experienced programmer, but new to the Ifc world.
I would like to use IfcOpenShell to create products with custom geometries programmatically, yielding in the end an IFC file with a scene of objects.
In this objective, I am now evaluating the options IfcOpenShell offers. I have found the example where the shape_builder can be used to create a custom geometry from simple elements such as extrusion (the table example).
Now, I am a bit lost in my new test : Importing a STEP geometry, getting the shape, and creating a representation from it, then linking it to an IfcProduct.
I must say I've tried a lot of combinations but I do not find useful information in the IfcOpenShell documentation.

My STEP geometry for testing is a sphere.

My questions :

1. Is STEP geometry fully compatible with IFC ? That means the shapes can be read from STEP and "copied" to IFC ?
2. Is my approach possible ? Basically this would allow me to import STEP files containing 3D models, and creating IfcProducts from them.
3. Given my code below, can you have a look and guide me ? I also added a debugging of the validation.
4. Overall, I am confused with the `run` method and the other methods called directly (for example, model.createIfcShapeRepresentation). It seems it's sometimes available in both options ? Which one is preferrable ?
5. Where to find the list of args I can pass to these methods ? I have args, kwargs, but no real helpful information. 
![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/18118854/02dadb29-e511-4e03-a529-a1c387b15063)

I know it's a lot of questions but it's a mess in my head these days, I am sure I am missing something, probably by laying out these interrogations it may be clear to you what's wrong with my approach.

The code 

```python
import os
import sys

import ifcopenshell
import ifcopenshell.api.geometry
import ifcopenshell.geom
import ifcopenshell.util.representation
import ifcopenshell.validate
from ifcopenshell.api import run
from OCC.Core.IFSelect import IFSelect_RetDone
from OCC.Core.STEPControl import STEPControl_Reader


def get_shape_from_step(fn):
    step_reader = STEPControl_Reader()
    status = step_reader.ReadFile(fn)
    if status == IFSelect_RetDone:
        step_reader.TransferRoot(1)
        if step_reader.NbShapes() >= 1:
            return step_reader.Shape(1)


def generate_readable_error(json_error):
    error_type = json_error.get("type", "Unknown")
    error_level = json_error.get("level", "Unknown")
    error_message = json_error.get("message", "No error message provided")
    error_attribute = json_error.get("attribute", "Unknown")
    instance = json_error.get("instance", "Unknown")

    readable_output = f"Error Type: {error_type}\n"
    readable_output += f"Error Level: {error_level}\n"
    readable_output += f"Error Message:\n    {error_message}\n"
    readable_output += f"Error Attribute: {error_attribute}\n"
    readable_output += f"Instance: {instance}"

    return readable_output


print(sys.version)
model = ifcopenshell.file()
print(ifcopenshell.version)
os.chdir(os.path.dirname(__file__))

# Create a blank model

# All projects must have one IFC Project element
project = run("root.create_entity", model, ifc_class="IfcProject", name="My Project")

# Geometry is optional in IFC, but because we want to use geometry in this example, let's define units
# Assigning without arguments defaults to metric units
run("unit.assign_unit", model)

# Let's create a modeling geometry context, so we can store 3D geometry (note: IFC supports 2D too!)
context = run("context.add_context", model, context_type="Model")

# In particular, in this example we want to store the 3D "body" geometry of objects, i.e. the body shape
body = run(
    "context.add_context",
    model,
    context_type="Model",
    context_identifier="Body",
    target_view="MODEL_VIEW",
    parent=context,
)

# Create a site, building, and storey. Many hierarchies are possible.
site = run("root.create_entity", model, ifc_class="IfcSite", name="My Site")
building = run("root.create_entity", model, ifc_class="IfcBuilding", name="Building A")
storey = run(
    "root.create_entity", model, ifc_class="IfcBuildingStorey", name="Ground Floor"
)

# Since the site is our top level location, assign it to the project
# Then place our building on the site, and our storey in the building
run("aggregate.assign_object", model, relating_object=project, product=site)
run("aggregate.assign_object", model, relating_object=site, product=building)
run("aggregate.assign_object", model, relating_object=building, product=storey)

# Create a body representation
body = ifcopenshell.util.representation.get_context(
    model, "Model", "Body", "MODEL_VIEW"
)

# Create a product
product = run("root.create_entity", model, ifc_class="IfcBuildingElementProxy")

# Obtain shape from step (TopoDS_Shape)
step_shape = get_shape_from_step("sphere.stp")

# Serialize to IFC (IfcProductDefinitionShape)
product_definition_shape = ifcopenshell.geom.serialise(model.schema, step_shape)

# First attempt to create a shape representation (does not work)
representation_type = "AdvancedBrep"
print(representation_type)
model.createIfcShapeRepresentation(
    ContextOfItems=body,
    RepresentationIdentifier="Body",
    RepresentationType=representation_type,
    Items=[product_definition_shape],
)

# Second attempt to create a shape representation (does not work)
run("representation.add_representation", model, product_definition_shape, body)

# Third attempt to create a shape representation (returns None)
representation = ifcopenshell.util.representation.get_representation(
    product_definition_shape,
    context="Model",
    subcontext="Body",
    target_view="MODEL_VIEW",
)

# Link the shape representation to the product
# First possible way ?
ifcopenshell.api.geometry.assign_representation(model,
    product=product,
    representation=representation)

# Second possible way ?
run(
    "geometry.assign_representation",
    model,
    product=product,
    representation=representation,
)

# Place our product in the ground floor
run("spatial.assign_container", model, relating_structure=storey, product=product)

# Write out to a file
model.write("product.ifc")

logger = ifcopenshell.validate.json_logger()
ifcopenshell.validate.validate(model, logger)

if logger.statements:
    for statement in logger.statements:
        print(generate_readable_error(statement))
        print("\n\n")
```

Thanks a lot for your help and time and all the best in the meantime.

Loïc

**Comment by aothms** [comment]
>     # First attempt to create a shape representation (does not work)

What did not work in this case?

> Is STEP geometry fully compatible with IFC ? That means the shapes can be read from STEP and "copied" to IFC ?

Not so easily. They share the same foundation and conceptually remain incredibly similar, but there's also deviations, sometimes subtle, sometimes rather large. E.g step uses multiple inheritance, in IFC this has been flattened to a single inheritance tree. IFC has added various parametric cross section profiles. These kind of differences.

So in your approach when you read the .stp file using opencascade and then serialize to ifc using ifcopenshell there is a reinterpretation happening twice.

But there is something else. AdvancedBrep representations have never really worked well in industry, because a lot of the implementations are mesh based so implementing the kind of surface parametrization required for AdvBrep is quite some effort, for little to gain, because shapes in AEC are typically so simple.

And then there is yet something else. IFC likes to think of itself as a semantic standard. So if you put an explicit brep in there it's not as descriptive as a procedural geometry (e.g an extrusion with boolean subtractions). This depends on the kind of element though. This is very much the case for e.g walls and columns, but less so for furniture.

> Overall, I am confused with the run method and the other methods called directly (for example, model.createIfcShapeRepresentation). It seems it's sometimes available in both options ? Which one is preferrable ?

Depends on preferences I guess. In many cases the high level api calls will take care of things you wouldn't easily think of your self such as representation contexts and what not. In other cases they are indeed roughly equivalent.

> Where to find the list of args I can pass to these methods ? I have args, kwargs, but no real helpful information.

The attributes are here https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/IfcShapeRepresentation.htm#8.18.3.17.3-Attributes Unfortunately the python code base is not really type aware. All help welcome here #1507 #3306

>     ifcopenshell.validate.validate(model, logger)

What output do you get actually?
